source/agora/node/Ledger.d

```diff
@@ -1535,12 +1578,17 @@ unittest
             auto cacheDB = new ManagedDatabase(":memory:");
             super(params, new Engine(TestStackMaxTotalSize, TestStackMaxItemSize),
                 new UTXOSet(stateDB),
-                new MemBlockStorage(blocks),
+                new MemBlockStorage(blocks ? blocks.takeExactly(1) : null),
                 new EnrollmentManager(stateDB, cacheDB, kp, params),
                 new TransactionPool(cacheDB),
                 new FeeManager(),
                 new MockClock(params.GenesisTimestamp +
                               (blocks.length * params.BlockInterval.total!"seconds")));
+            if (blocks)
+            {
+                blocks[1 .. $].map!(b => b.header.height).each!(h => simulatePreimages(this.log, enroll_man, genesis_validator_keys, h));
+                blocks[1 .. $].each!(b => this.addValidatedBlock(b));
+            }
```

Comment:
Instead of patching it here, we should feed the `ThrowingLedger` the right blocks (valid ones).

Reply:
I don't see how we can store the `pre-images`  without an `EnrollmentManager` which is a member of the `Ledger`. That is why I patch the blocks here in the constructor.

Reply:
On second thought I could add `EnrollmentManager` as a parameter to `TestLedger` and `ThrowingLedger`. It would also mean that a few of the existing parameters could be removed as they are members of `EnrollmentManager`.

Reply:
Why not instantiate it in the ctor, populate the pre-images, and then build the ledger ?
My main concern is using a different road to add the block instead of using the code that is already doing this in the Ledger.

Reply:
Do you mean using `acceptBlock` to add blocks rather than `addValidatedBlock` which is what I use here?

Reply:
I just tried replacing `addValidatedBlock` with `acceptBlock` and the Ledger tests pass still. That means it is now using the same code to add the blocks. Just inside the constructor here rather than before, which will involve updating a lot of calling code.

Reply:
I mean that L1581 (previously L1538) should not be modified (`blocks.takeExactly(1)`).

The Ledger has recovery code in the constructor: https://github.com/bosagora/agora/blob/03a97193bc93d0cc80167e0e85892e635d3b62a5/source/agora/node/Ledger.d#L182-L220

What L1581 does is to *completely bypass that code*, then replicate its functionality here. My guess as for why this is being done is that the EM does not have the pre-images. One way to work around this is by adding them ahead of time, which require instantiating the EM, adding the pre-images, then passing it to Ledger's ctor.
However that might currently only work with 1 consensus round, so further changes might be necessary. But *bypassing a piece of code just to duplicate it in unittest* is going to come back to bite us back very soon.

Reply:
ok now I see what you mean, thanks for the explanation. I will add the pre-images to the EM in the `TestLedger` and `ThrowingLedger` before calling the parent constructor.

Reply:
Has been refactored.